### PR TITLE
OSDOCS#9533:Updates for preflight cluster upgrades for ROSA Classic.

### PR DESCRIPTION
**Pull request by AedinC**
**New PR created (#78825) that will take over these updates.**
 
Version(s):
4.16+

Issue:
[OSDOCS-9533](https://issues.redhat.com/browse/OSDOCS-9533)

Link to docs preview:
[Upgrading ROSA Classic clusters](https://76632--ocpdocs-pr.netlify.app/openshift-rosa/latest/upgrading/rosa-upgrading-sts.html)

Peer review:
- [x] Peer reviewer has approved this change.
 
SME review:
- [ ] SME has approved this change. 

QE review:
- [ ] QE has approved this change.

Additional information:
**Note for reviewers**: We are in the process of renaming ROSA classic to ROSA (classic architecture) and ROSA with HCP to ROSA. This means we must change the variables for both variables. We have not yet done this, so throughout this PR, I have used ROSA (classic architecture) as a placeholder until we have the new variables.

### CODE REVIEW

File: modules/rosa-how-upgrades-work.adoc [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+:_mod-docs-content-type: CONCEPT
+[id="rosa-how-upgrades-work_{context}"]
+= How ROSA cluster upgrades work
+
+Upgrades are manually initiated (one-time) or automatically scheduled (recurring). Red Hat Site Reliability Engineers (SREs) monitor upgrade progress and either proactively notify to take any corrective actions or remedy any issues encountered.
+
+[id="rosa-version-operator_{context}"]
+== Cluster Version Operator
+
+The `Cluster Version` Operator (CVO) is the primary component that orchestrates and facilitates the OpenShift Container Platform update process. The scheduling, monitoring, and notifications of ROSA cluster upgrades are orchestrated through the `Managed Upgrade` Operator (MUO). The MUO manages the orchestration of automated in-place cluster upgrades by satisfying the operating conditions that a managed cluster must hold, both pre- and post-invocation of the cluster upgrade.
```
Comment (bchandra-ocp) [inline modules/rosa-how-upgrades-work.adoc:16]: We're covering MUO under the title CVO. Perhaps we separate them out or perhaps leave out the title and have it just under 'how rosa cluster upgrades work' because the content is small

File: modules/rosa-how-upgrades-work.adoc [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+====
++
+. During the upgrade, the control plane components are updated to the new version.
+. Next, individual cluster Operators perform update tasks on their domain of the cluster.
+. Finally, the `Machine Config` Operator (MCO) begins updating the system configuration and operating system of every node. During this step, each node is rebooted after successfully draining the workloads running on the node.
+.. During the update of each node, workloads are drained honoring the PDBs. Workloads with PDBs that do not allow disruptions essentially block the draining of the node, increasing the elapsed time for the cluster update.
+.. During the update of every node in the cluster, the cluster update waits until the time specified by the node drain grace period to allow for safely draining the workloads. Upon reaching the node drain grace period, the node is force drained to allow for cluster upgrade to progress. You can only configure the node drain grace period at cluster level before initiating the upgrade and you cannot change it after the cluster upgrade begins.
```
Comment (bchandra-ocp) [inline modules/rosa-how-upgrades-work.adoc:44]: suggest emphasizing this "node drain grace period" terminology through bold letters if possible.

**Review by bchandra-ocp — CHANGES_REQUESTED** [review]
The "Upgrading a ROSA Classic clusters" is the main title. there's a sub title "Upgrading a ROSA Classic cluster". Can we reword this to avoid duplication?

File: upgrading/rosa-upgrading-sts.adoc [hunk 3]
```diff
@@ -11,21 +11,32 @@ toc::[]
...
-* Individual upgrades through the {cluster-manager} console
+You must update {product-title} (ROSA) Classic clusters using the ROSA CLI (`rosa`) or the {cluster-manager} console.
 
 [NOTE]
 ====
-When you follow a scheduled upgrade policy, a delay of an hour or more before the upgrade process begins might occur, even if the upgrade is configured to begin immediately. Additionally, the duration of the upgrade might vary based on your workload configuration.
+The actual start time of the cluster upgrade will be within one hour of the upgrade schedule time. Additionally, the duration of the upgrade might vary based on your workload configuration.
 ====
 
+When a ROSA Classic cluster that uses AWS Security Token Services (STS) is upgraded, the ROSA CLI verifies the account and operator role policies for the chosen cluster are compatible with the target version of the upgrade. If the policies are compatible, the CLI automatically upgrades the IAM policies. If the policies are not compatible with the chosen upgrade version, the CLI prompts you to upgrade them. When scheduling the upgrade, you provide administrative acknowledgment to automatically upgrade these policies, if required.
```
Comment (bchandra-ocp) [inline upgrading/rosa-upgrading-sts.adoc:34]: the sentence "If the policies are compatible, the CLI automatically upgrades the cluster"

Comment (bchandra-ocp) [reply upgrading/rosa-upgrading-sts.adoc:34]: "When scheduling the upgrade, you provide administrative acknowledgment to confirm you have reviewed the changes involved with the upgrade, if required."

File: modules/rosa-upgrading-cli-tutorial.adoc [hunk 4]
```diff
@@ -58,9 +57,51 @@ $ rosa upgrade cluster --cluster=<cluster_name|cluster_id> --control-plane
 ----
 endif::rosa-hcp[]
 +
-The cluster is scheduled for an immediate upgrade. This action can take an hour or longer, depending on your workload configuration, such as pod disruption budgets.
```
Comment (bchandra-ocp) [inline modules/rosa-upgrading-cli-tutorial.adoc:61]: The cluster is scheduled for an immediate upgrade as denoted by the Scheduled Time. The upgrade will begin within one hour from the Scheduled Time.

**Review by bchandra-ocp — CHANGES_REQUESTED** [review]
thanks for working on this, provided some feedback.

**Comment by bchandra-ocp** [comment]
@AedinC sorry for getting back late on this. Can you also please make sure SME (Ravi) has reviewed as well.

File: modules/rosa-how-upgrades-work.adoc (hunk 2 above)
Comment (AedinC) [reply modules/rosa-how-upgrades-work.adoc:41]: As per the [RH SSG](https://redhat-documentation.github.io/supplementary-style-guide/#user-interface-elements), we use bold text for all graphical user interface (GUI) element names, and not to emphasize terms. I can italicize the first instance of 'node drain grace period' to emphasize the first appearance of the new term.

File: upgrading/rosa-upgrading-sts.adoc [hunk 5]
```diff
@@ -1,31 +1,48 @@
...
 include::_attributes/attributes-openshift-dedicated.adoc[]
 :context: rosa-upgrading-sts
 
 toc::[]
 
+Use one of the following methods to upgrade ROSA (classic architecture) clusters:
+
+* Manually via the ROSA CLI (`rosa`) - Start a one-time immediate upgrade or schedule a one-time upgrade for a future date or time.
+* Manually via the {cluster-manager} UI - Start a one-time immediate upgrade or schedule a one-time upgrade for a future date or time.
+* Automated upgrades - Set an upgrade window for recurring y-stream upgrades whenever a new version is available without needing to manually schedule it. Minor versions have to be manually scheduled.
```
Comment (bscott-rh) [inline upgrading/rosa-upgrading-sts.adoc:13]:
```suggestion
* Automated upgrades - Set an upgrade window for automatic recurring y-stream upgrades whenever a new version is available. Upgrades to minor versions must be manually scheduled.
```
Reworded for clarity

File: upgrading/rosa-upgrading-sts.adoc [hunk 6]
```diff
@@ -1,31 +1,48 @@
...
+
 [id="rosa-lifecycle-policy_{context}"]
 == Life cycle policies and planning
 
 To plan an upgrade, review the xref:../rosa_architecture/rosa_policy_service_definition/rosa-life-cycle.adoc#rosa-life-cycle[{product-title} update life cycle]. The life cycle page includes release definitions, support and upgrade requirements, installation policy information and life cycle dates.
 
-Upgrades are manually initiated or automatically scheduled. Red Hat Site Reliability Engineers (SREs) monitor upgrade progress and remedy any issues encountered.
+OpenShift update channels are the mechanism that you can use to decide which Red Hat OpenShift Container Platform minor version to update your clusters to. {product-title} only supports updates through the `stable` channel. To learn more about OpenShift update channels and releases, see link:https://docs.openshift.com/container-platform/4.15/updating/understanding_updates/understanding-update-channels-release.html[Understanding update channels and releases].
```
Comment (bscott-rh) [inline upgrading/rosa-upgrading-sts.adoc:20]:
```suggestion
You may use update channels to decide which Red Hat OpenShift Container Platform minor version to update your clusters to. {product-title} only supports updates through the `stable` channel. To learn more about OpenShift update channels and releases, see link:https://docs.openshift.com/container-platform/4.15/updating/understanding_updates/understanding-update-channels-release.html[Understanding update channels and releases].
```

**Review by bscott-rh — COMMENTED** [review]
Nice work! I left some rewording suggestions, typo fixes and syntax recommendations. Let me know if you have any questions.

File: modules/rosa-deleting-cluster-upgrade-cli.adoc [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+----
++
+.Example output
+[source,terminal]
+----
+? Are you sure you want to cancel scheduled upgrade on cluster my-cluster? Yes
+I: Successfully canceled scheduled upgrade on cluster 'my-cluster'
+----
+
+You will receive a notification confirming that the scheduled upgrade has been canceled.
```
[the remark this answers is not in the document]
Comment (AedinC) [reply modules/rosa-deleting-cluster-upgrade-cli.adoc:41]: Yes good call- it's by email. I'll add that now, thanks.

File: modules/rosa-how-upgrades-work.adoc [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+. Next, individual cluster Operators perform update tasks on their domain of the cluster.
+. Finally, the `Machine Config` Operator (MCO) begins updating the system configuration and operating system of every node. During this step, each node is rebooted after successfully draining the workloads running on the node.
+.. During the update of each node, workloads are drained honoring the PDBs. Workloads with PDBs that do not allow disruptions essentially block the draining of the node, increasing the elapsed time for the cluster update.
+.. During the update of every node in the cluster, the cluster update waits until the time specified by the _node drain grace period_ to allow for safely draining the workloads. Upon reaching the node drain grace period, the node is force drained to allow for cluster upgrade to progress. You can only configure the node drain grace period at cluster level before initiating the upgrade and you cannot change it after the cluster upgrade begins.
+.. When nodes of the cluster are updated, the `Machine Config` Operator selects one node per machine config pool at a time according to their age, starting with the oldest.
```
[the remark this answers is not in the document]
Comment (wking) [reply modules/rosa-how-upgrades-work.adoc:42]: We might not want to commit to one-at-a-time here, with [the public XCMSTRAT-187][1] mulling over making MachineConfigPool `maxUnavailable` an OCM-configurable knob 

[1]: https://issues.redhat.com/browse/XCMSTRAT-187

File: modules/rosa-upgrading-cli-tutorial.adoc [hunk 9]
```diff
@@ -40,14 +39,14 @@ $ rosa describe cluster --cluster=<cluster_name|cluster_id> <1>
 $ rosa list upgrade --cluster=<cluster_name|cluster_id>
 ----
 +
-The command returns a list of versions to which the cluster can be upgraded, including a recommended version.
+The command returns a list of versions to which the cluster can be upgraded, including a recommended version. The recommendation is based on the conditional update risks. Each known risk might apply to all clusters or only clusters matching certain conditions. Refer to the OpenShift release notes to evaluate, validate and determine the appropriate version to upgrade to.
```
[the remark this answers is not in the document]
Comment (AedinC) [reply modules/rosa-upgrading-cli-tutorial.adoc:42]: Yes, it should be the OCP release notes.

**Comment by bscott-rh** [comment]
Peer review LGTM.

File: modules/rosa-how-upgrades-work.adoc [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+Upgrades are manually initiated (one-time) or automatically scheduled (recurring). Red{nbsp}Hat Site Reliability Engineers (SREs) monitor upgrade progress and either proactively notify you to take corrective actions or remedy issues encountered.
+
+The Cluster Version Operator (CVO) is the primary component that orchestrates and facilitates the OpenShift Container Platform update process. The scheduling, monitoring, and notifications of ROSA (classic architecture) cluster upgrades are orchestrated through the Managed Upgrade Operator (MUO). The MUO manages the orchestration of automated in-place cluster upgrades by satisfying the operating conditions that a managed cluster must hold before and after initiating the cluster upgrade.
+
+[id="rosa-upgrade-scheduled-time_{context}"]
+== Cluster upgrade scheduled time
+
+You can schedule cluster upgrades by setting the scheduled time. This is when the preparation for the cluster upgrade begins with pre-upgrade health checks and additional compute capacity creation. The actual cluster upgrade starts within one hour from the scheduled time. You receive an email notification when the cluster upgrade starts.
```
Comment (ravitri) [inline modules/rosa-how-upgrades-work.adoc:18]: The Pre-Health Check (PHC) is run in following two scenarios:
- If an upgrade is scheduled more than 2 hours from now, PHC is run and user is notified if there are any failures. This PHC is in the "New" phase of upgrade.
- When the upgrade is "immediate" or within 2 hours, PHC is run just before the upgrade begins. This PHC is in the "Upgrading" phase of upgrade.

This means that PHC is always run atleast once during upgrading phase but can be run additionally in advance too if the upgrade is scheduled more than 2 hours from now.

File: modules/rosa-how-upgrades-work.adoc [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+You can observe the status of the cluster upgrade by running the `oc get clusterversion` command in the ROSA CLI (`rosa`).
+
+[id="rosa-cluster-upgrade-overview_{context}"]
+== ROSA (classic architecture) upgrade overview
+
+The following high-level steps that occur during the ROSA (classic architecture)cluster update:
+
+. Before the cluster upgrade begins, a cluster health check is performed by the  MUO. If the MUO identifies an issue that requires corrective action, you will be notified. Some examples of the cluster health checks that MUO performs include:
```
Comment (ravitri) [inline modules/rosa-how-upgrades-work.adoc:27]: We can add one more high level step for when upgrade is scheduled in advance which would trigger the PreHealthCheck and notify user of any failures which can be addressed before the upgrade begins.

File: modules/rosa-how-upgrades-work.adoc [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+You can schedule cluster upgrades by setting the scheduled time. This is when the preparation for the cluster upgrade begins with pre-upgrade health checks and additional compute capacity creation. The actual cluster upgrade starts within one hour from the scheduled time. You receive an email notification when the cluster upgrade starts.
+
+You can observe the status of the cluster upgrade by running the `oc get clusterversion` command in the ROSA CLI (`rosa`).
+
+[id="rosa-cluster-upgrade-overview_{context}"]
+== ROSA (classic architecture) upgrade overview
+
+The following high-level steps that occur during the ROSA (classic architecture)cluster update:
+
```
Comment (ravitri) [inline modules/rosa-how-upgrades-work.adoc:26]: We can add one more high level step for when upgrade is scheduled in advance which would trigger the PreHealthCheck and notify user of any failures which can be addressed before the upgrade begins.

File: modules/rosa-how-upgrades-work.adoc [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+. Before the cluster upgrade begins, a cluster health check is performed by the  MUO. If the MUO identifies an issue that requires corrective action, you will be notified. Some examples of the cluster health checks that MUO performs include:
+** Identifying any Pod Disruption Budgets (PDBs) that can potentially block or delay the update of the nodes.
+** Ensuring cluster Operators are available and healthy.
+** Discovering any nodes that are not in a `Ready` status.
+. A temporary compute node is created in the cluster to allow for the scheduling of drained pods during the update.
```
[the remark this answers is not in the document]
Comment (AedinC) [reply modules/rosa-how-upgrades-work.adoc:31]: Thanks @ravitri - how does it not always happen? Is it up to the user to select this? Or how does it work?

File: modules/rosa-how-upgrades-work.adoc [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+// Module included in the following assemblies:
+//
+// upgrading/rosa-upgrading-sts.adoc
+
+
+:_mod-docs-content-type: CONCEPT
+[id="rosa-how-upgrades-work_{context}"]
+= How ROSA (classic architecture) cluster upgrades work
+
+Upgrades are manually initiated (one-time) or automatically scheduled (recurring). Red Hat Site Reliability Engineers (SREs) monitor upgrade progress and either proactively notify you to take corrective actions or remedy issues encountered.
```
[the remark this answers is not in the document]
Comment (AedinC) [reply modules/rosa-how-upgrades-work.adoc]: @ravitri Can you advise if I can add anything here that would make things clearer for cluster admins? Thanks!

[the remark this answers is not in the document]
Comment (AedinC) [reply modules/rosa-how-upgrades-work.adoc]: The scope of this PR is to cover the overall theory and working of the PHC feature and MUO. There will be a KCS article which will be dedicated for troubleshooting and diagnosis, and will cover issues that customers can fix.
